File: sdk/dotnet/src/Snaploom.Capture/CaptureClient.cs
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text;

namespace Snaploom.Capture;

/// <summary>A disposable .NET 8 client for the separately installed Snaploom Capture Host.</summary>
public sealed unsafe class CaptureClient : IDisposable
{
    private readonly object _gate = new();
    private readonly INativeApi _native;
    private SafeCaptureClientHandle? _handle;

    /// <summary>Creates a client using the native asset selected from the current package RID.</summary>
    public CaptureClient(CaptureClientOptions? options = null)
        : this(NativeApi.Instance, options, enforcePlatform: true)
    {
    }

    internal CaptureClient(INativeApi native, CaptureClientOptions? options = null, bool enforcePlatform = false)
    {
        ArgumentNullException.ThrowIfNull(native);
        if (enforcePlatform)
        {
            PlatformSupport.EnsureSupported();
        }

        _native = native;
        RuntimeVersion = QueryRuntimeVersion(native);
        _handle = CreateHandle(native, options);
    }

    /// <summary>The ABI and semantic version reported by the loaded native SDK.</summary>
    public CaptureSdkVersion RuntimeVersion { get; }

    internal static int ActiveRequestHandleCount => RequestState.ActiveHandleCount;

    /// <summary>Starts one capture and completes only after the native terminal callback.</summary>
    public Task<CaptureResult> CaptureAsync(
        CaptureOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        SafeCaptureClientHandle handle;
        var addRef = false;
        lock (_gate)
        {
            ObjectDisposedException.ThrowIf(_handle is null, this);
            handle = _handle;
            handle.DangerousAddRef(ref addRef);
        }

        var request = new RequestState(_native, handle, cancellationToken);
        var userData = request.AllocateHandle();
        var nativeOptions = CreateOptions(options);
        ulong requestId = 0;

        try
        {
            var status = cancellationToken.CanBeCanceled
                ? _native.Start(handle, &nativeOptions, &CompletionCallback, userData, &requestId)
                : _native.Start(handle, &nativeOptions, &CompletionCallback, userData, null);
            if (status != NativeConstants.StatusOk)
            {
                request.AbortStart();
                ThrowStatus(status, nameof(options));
            }

            request.PublishRequestId(requestId);
            return request.Task;
        }
        catch
        {
            request.AbortStart();
            throw;
        }
        finally
        {
            if (addRef)
            {
                handle.DangerousRelease();
            }
        }
    }

    /// <summary>Closes the native client. This may block while accepted callbacks finish.</summary>
    public void Dispose()
    {
        SafeCaptureClientHandle? handle;
        lock (_gate)
        {
            handle = _handle;
            _handle = null;
        }

        handle?.Dispose();
    }

    [UnmanagedCallersOnly(CallConvs = [typeof(CallConvCdecl)])]
    private static void CompletionCallback(nint client, nint completion, nint userData)
    {
        _ = client;
        RequestState? request = null;
        try
        {
            var gcHandle = GCHandle.FromIntPtr(userData);
            request = gcHandle.Target as RequestState;
            if (request is null)
            {
                return;
            }

            request.Complete(completion);
        }
        catch (Exception error)
        {
            request?.FaultFromCallback(error);
        }
        finally
        {
            request?.ReleaseAfterCallback();
        }
    }

    private static CaptureSdkVersion QueryRuntimeVersion(INativeApi native)
    {
        var version = new NativeVersionInfo { StructSize = (uint)sizeof(NativeVersionInfo) };
        var status = native.Version(ref version);
        if (status != NativeConstants.StatusOk)
        {
            ThrowStatus(status, "version");
        }

        if (version.StructSize < sizeof(NativeVersionInfo)
            || version.AbiMajor != NativeConstants.AbiMajor
            || version.SdkSemver.Data is null
            || version.SdkSemver.Length is 0 or > 256)
        {
            throw new CaptureSdkException("The loaded Snaploom Capture native SDK has an incompatible ABI.");
        }

        var semver = Encoding.UTF8.GetString(version.SdkSemver.Data, checked((int)version.SdkSemver.Length));
        var separator = semver.IndexOf('.');
        if (separator <= 0
            || !uint.TryParse(semver.AsSpan(0, separator), out var semverMajor)
            || semverMajor != NativeConstants.PackageSemverMajor)
        {
            throw new CaptureSdkException("The loaded Snaploom Capture native SDK has an incompatible semantic version.");
        }

        return new CaptureSdkVersion(version.AbiMajor, semver);
    }

    private static SafeCaptureClientHandle CreateHandle(INativeApi native, CaptureClientOptions? options)
    {
        var hostPath = options?.HostExecutableOverride;
        if (!string.IsNullOrEmpty(hostPath) && !Path.IsPathFullyQualified(hostPath))
        {
            throw new ArgumentException("The Capture Host override must be an absolute path.", nameof(options));
        }

        var hostBytes = string.IsNullOrEmpty(hostPath) ? Array.Empty<byte>() : Encoding.UTF8.GetBytes(hostPath);
        fixed (byte* hostPointer = hostBytes)
        {
            var config = new NativeClientConfig
            {
                StructSize = (uint)sizeof(NativeClientConfig),
                HostExecutableOverride = new NativeUtf8View
                {
                    Data = hostBytes.Length == 0 ? null : hostPointer,
                    Length = (ulong)hostBytes.Length,
                },
                LaunchTimeoutMilliseconds = ToUInt32Milliseconds(options?.LaunchTimeout, nameof(options.LaunchTimeout)),
                HandshakeTimeoutMilliseconds = ToUInt32Milliseconds(options?.HandshakeTimeout, nameof(options.HandshakeTimeout)),
            };
            var status = native.Create(&config, out var client);
            if (status != NativeConstants.StatusOk)
            {
                ThrowStatus(status, nameof(options));
            }

            if (client == 0)
            {
                throw new CaptureSdkException("The native SDK returned a null client after a successful create.");
            }

            return new SafeCaptureClientHandle(client, native);
        }
    }

    private static NativeCaptureOptions CreateOptions(CaptureOptions? options)
    {
        return new NativeCaptureOptions
        {
            StructSize = (uint)sizeof(NativeCaptureOptions),
            Flags = options?.DisableClipboard == true ? NativeConstants.DisableClipboard : 0,
            InteractionTimeoutMilliseconds = ToUInt64Milliseconds(options?.InteractionTimeout, nameof(options.InteractionTimeout)),
        };
    }

    private static uint ToUInt32Milliseconds(TimeSpan? value, string parameter)
    {
        var milliseconds = ToUInt64Milliseconds(value, parameter);
        if (milliseconds > uint.MaxValue)
        {
            throw new ArgumentOutOfRangeException(parameter);
        }

        return (uint)milliseconds;
    }

    private static ulong ToUInt64Milliseconds(TimeSpan? value, string parameter)
    {
        if (value is null)
        {
            return 0;
        }

        if (value <= TimeSpan.Zero || value.Value.TotalMilliseconds > ulong.MaxValue)
        {
            throw new ArgumentOutOfRangeException(parameter);
        }

        return checked((ulong)Math.Ceiling(value.Value.TotalMilliseconds));
    }

    private static void ThrowStatus(uint status, string parameter)
    {
        switch (status)
        {
            case NativeConstants.StatusInvalidArgument:
            case NativeConstants.StatusInvalidStructSize:
                throw new CaptureArgumentException(status);
            case NativeConstants.StatusClientClosed:
                throw new ObjectDisposedException("CaptureClient", $"The native client is closed (status {status}).");
            case NativeConstants.StatusOutOfMemory:
                throw new OutOfMemoryException("The native Snaploom Capture SDK could not allocate memory.");
            default:
                throw new CaptureSdkException($"Snaploom Capture native call '{parameter}' failed with status {status}.", status);
        }
    }

    private sealed class RequestState
    {
        private readonly object _gate = new();
        private readonly INativeApi _native;
        private readonly SafeCaptureClientHandle _client;
        private readonly CancellationToken _token;
        private readonly TaskCompletionSource<CaptureResult> _completion =
            new(TaskCreationOptions.RunContinuationsAsynchronously);
        private CancellationTokenRegistration _registration;
        private nint _userData;
        private ulong _requestId;
        private bool _requestIdPublished;
        private bool _cancelRequested;
        private bool _terminal;

        internal RequestState(INativeApi native, SafeCaptureClientHandle client, CancellationToken token)
        {
            _native = native;
            _client = client;
            _token = token;
            if (token.CanBeCanceled)
            {
                _registration = token.Register(static state => ((RequestState)state!).RequestCancel(), this);
            }
        }

        internal static int ActiveHandleCount;

        internal Task<CaptureResult> Task => _completion.Task;

        internal nint AllocateHandle()
        {
            var handle = GCHandle.Alloc(this, GCHandleType.Normal);
            var userData = GCHandle.ToIntPtr(handle);
            _userData = userData;
            Interlocked.Increment(ref ActiveHandleCount);
            return userData;
        }

        internal void PublishRequestId(ulong requestId)
        {
            var shouldCancel = false;
            lock (_gate)
            {
                if (_terminal)
                {
                    return;
                }

                _requestId = requestId;
                _requestIdPublished = true;
                shouldCancel = _cancelRequested;
            }

            if (shouldCancel)
            {
                SendCancel(requestId);
            }
        }

        internal void Complete(nint completionPointer)
        {
            try
            {
                if (completionPointer == 0)
                {
                    throw new CaptureSdkException("The native SDK returned a null completion.");
                }

                var completion = *(NativeCompletion*)completionPointer;
                CompleteTask(CopyCompletion(completion));
            }
            catch (OperationCanceledException) when (_token.IsCancellationRequested)
            {
                _completion.TrySetCanceled(_token);
            }
            catch (Exception error)
            {
                _completion.TrySetException(error);
            }
            finally
            {
                _native.FreeCompletion(completionPointer);
            }
        }

        internal void FaultFromCallback(Exception error)
        {
            _completion.TrySetException(new CaptureSdkException("A managed callback bridge failure occurred.", innerException: error));
        }

        internal void ReleaseAfterCallback()
        {
            lock (_gate)
            {
                _terminal = true;
            }

            _registration.Unregister();
            FreeHandleOnce();
        }

        internal void AbortStart()
        {
            lock (_gate)
            {
                if (_terminal)
                {
                    return;
                }

                _terminal = true;
            }

            _registration.Unregister();
            FreeHandleOnce();
        }

        private void RequestCancel()
        {
            ulong requestId = 0;
            var shouldCancel = false;
            lock (_gate)
            {
                if (_terminal)
                {
                    return;
                }

                _cancelRequested = true;
                if (_requestIdPublished)
                {
                    requestId = _requestId;
                    shouldCancel = true;
                }
            }

            if (shouldCancel)
            {
                SendCancel(requestId);
            }
        }

        private void SendCancel(ulong requestId)
        {
            var addRef = false;
            try
            {
                _client.DangerousAddRef(ref addRef);
                _ = _native.Cancel(_client, requestId);
            }
            catch (ObjectDisposedException)
            {
                // Destroy records cancellation for every accepted request and waits for terminal callbacks.
            }
            finally
            {
                if (addRef)
                {
                    _client.DangerousRelease();
                }
            }
        }

        private CaptureResult CopyCompletion(NativeCompletion completion)
        {
            if (completion.StructSize < sizeof(NativeCompletion)
                || completion.Reserved0 != 0
                || completion.Reserved1 != 0
                || completion.Reserved2 != 0
                || completion.Reserved3 != 0
                || (completion.Flags & ~(NativeConstants.ClipboardWritten | NativeConstants.Retryable)) != 0)
            {
                throw new CaptureSdkException("The native SDK returned an invalid completion layout.");
            }

            if (completion.Kind == NativeConstants.CompletionCompleted)
            {
                if (completion.ErrorCode != 0
                    || completion.PngData == 0
                    || completion.PngSize == 0
                    || completion.PngSize > NativeConstants.MaximumPngBytes
                    || completion.PngSize > int.MaxValue
                    || completion.PixelWidth == 0
                    || completion.PixelHeight == 0
                    || (completion.Flags & NativeConstants.Retryable) != 0)
                {
                    throw new CaptureSdkException("The native SDK returned an invalid completed capture.");
                }

                var png = new byte[checked((int)completion.PngSize)];
                Marshal.Copy(completion.PngData, png, 0, png.Length);
                return new CaptureResult(
                    png,
                    completion.PixelWidth,
                    completion.PixelHeight,
                    (completion.Flags & NativeConstants.ClipboardWritten) != 0);
            }

            if (completion.PngData != 0
                || completion.PngSize != 0
                || completion.PixelWidth != 0
                || completion.PixelHeight != 0
                || (completion.Flags & NativeConstants.ClipboardWritten) != 0)
            {
                throw new CaptureSdkException("The native SDK returned image data for a non-completed terminal.");
            }

            if (completion.Kind == NativeConstants.CompletionCanceled && completion.ErrorCode == 0 && completion.Flags == 0)
            {
                if (_token.IsCancellationRequested)
                {
                    throw new OperationCanceledException(_token);
                }

                throw new CaptureCanceledException();
            }

            if (completion.Kind == NativeConstants.CompletionFailed
                && completion.ErrorCode != 0
                && (completion.Flags & ~NativeConstants.Retryable) == 0)
            {
                var namePointer = _native.ErrorName(completion.ErrorCode);
                var name = namePointer == 0 ? "UNKNOWN" : Marshal.PtrToStringUTF8(namePointer) ?? "UNKNOWN";
                throw new CaptureException(
                    (CaptureErrorCode)completion.ErrorCode,
                    (completion.Flags & NativeConstants.Retryable) != 0,
                    name);
            }

            throw new CaptureSdkException($"The native SDK returned an unknown completion kind {completion.Kind}.");
        }

        private void CompleteTask(CaptureResult result)
        {
            _completion.TrySetResult(result);
        }

        private void FreeHandleOnce()
        {
            var userData = Interlocked.Exchange(ref _userData, 0);
            if (userData == 0)
            {
                return;
            }

            GCHandle.FromIntPtr(userData).Free();
            Interlocked.Decrement(ref ActiveHandleCount);
        }
    }
}

internal static class PlatformSupport
{
    internal const string UnsupportedMessage =
        "SNAPLOOM001: Snaploom.Capture supports only win-x64 and osx-arm64.";

    internal static void EnsureSupported()
    {
        var architecture = RuntimeInformation.ProcessArchitecture;
        if ((OperatingSystem.IsWindows() && architecture == Architecture.X64)
            || (OperatingSystem.IsMacOS() && architecture == Architecture.Arm64))
        {
            return;
        }

        throw new PlatformNotSupportedException(UnsupportedMessage);
    }
}

File: sdk/swift/Sources/SnaploomCapture/CaptureClient.swift
import CSnaploomCapture
import Foundation

/// A Swift concurrency wrapper for the separately installed Snaploom Capture Host.
public final class CaptureClient: @unchecked Sendable {
  private let state: ClientState

  /// The ABI and semantic version reported by the loaded native SDK.
  public let runtimeVersion: CaptureSDKVersion

  public convenience init(options: CaptureClientOptions = .init()) throws {
    try self.init(native: SystemNativeCaptureAPI(), options: options)
  }

  internal init(
    native: any NativeCaptureAPI,
    options: CaptureClientOptions = .init()
  ) throws {
    runtimeVersion = try Self.queryRuntimeVersion(native: native)
    let client = try Self.createClient(native: native, options: options)
    state = ClientState(native: native, client: client)
  }

  deinit {
    try? state.close()
  }

  /// Starts one capture and resumes only after the native terminal callback.
  public func capture(options: CaptureOptions = .init()) async throws -> CaptureResult {
    let box = RequestBox(native: state.native) { [state] requestID in
      state.requestCancel(requestID: requestID)
    }

    return try await withTaskCancellationHandler {
      try await withCheckedThrowingContinuation {
        (continuation: CheckedContinuation<CaptureResult, any Error>) in
        box.install(continuation: continuation)
        let retained = Unmanaged.passRetained(box)
        let userData = retained.toOpaque()
        var nativeOptions = snaploom_capture_options_v1()
        nativeOptions.struct_size = UInt32(
          MemoryLayout<snaploom_capture_options_v1>.size
        )
        nativeOptions.flags =
          options.disableClipboard
          ? NativeConstants.disableClipboard
          : 0
        nativeOptions.interaction_timeout_ms =
          options.interactionTimeoutMilliseconds ?? 0
        var requestID: UInt64 = 0
        let status = state.start(
          options: &nativeOptions,
          callback: captureCallback,
          userData: userData,
          requestID: &requestID
        )
        if status == NativeConstants.statusOK {
          box.publish(requestID: requestID)
        } else {
          retained.release()
          box.failStart(error: Self.error(for: status))
        }
      }
    } onCancel: {
      box.requestCancel()
    }
  }

  /// Closes the client. This may block while accepted native callbacks finish.
  public func close() throws {
    try state.close()
  }

  private static func queryRuntimeVersion(
    native: any NativeCaptureAPI
  ) throws -> CaptureSDKVersion {
    var version = snaploom_capture_version_info_v1()
    version.struct_size = UInt32(
      MemoryLayout<snaploom_capture_version_info_v1>.size
    )
    let status = native.version(&version)
    guard status == NativeConstants.statusOK else {
      throw error(for: status)
    }
    guard
      version.struct_size >= MemoryLayout<snaploom_capture_version_info_v1>.size,
      version.abi_major == NativeConstants.abiMajor,
      let data = version.sdk_semver.data,
      version.sdk_semver.length > 0,
      version.sdk_semver.length <= 256,
      let count = Int(exactly: version.sdk_semver.length)
    else {
      throw CaptureError.incompatibleNative(reason: "ABI version information is invalid")
    }
    let semver = String(decoding: UnsafeBufferPointer(start: data, count: count), as: UTF8.self)
    guard
      let majorText = semver.split(separator: ".", maxSplits: 1).first,
      let major = UInt32(majorText),
      major == NativeConstants.packageSemverMajor
    else {
      throw CaptureError.incompatibleNative(reason: "native package major is incompatible")
    }
    return CaptureSDKVersion(abiMajor: version.abi_major, nativeSemver: semver)
  }

  private static func createClient(
    native: any NativeCaptureAPI,
    options: CaptureClientOptions
  ) throws -> OpaquePointer {
    if let path = options.hostExecutableOverride,
      !(path as NSString).isAbsolutePath
    {
      throw CaptureError.invalidArgument(status: NativeConstants.statusInvalidArgument)
    }
    var config = snaploom_capture_client_config_v1()
    config.struct_size = UInt32(
      MemoryLayout<snaploom_capture_client_config_v1>.size
    )
    config.launch_timeout_ms = options.launchTimeoutMilliseconds ?? 0
    config.handshake_timeout_ms = options.handshakeTimeoutMilliseconds ?? 0
    let hostBytes = Array((options.hostExecutableOverride ?? "").utf8)
    var client: OpaquePointer?
    let status = hostBytes.withUnsafeBufferPointer { bytes in
      config.host_executable_override.data = bytes.isEmpty ? nil : bytes.baseAddress
      config.host_executable_override.length = UInt64(bytes.count)
      return native.create(&config, &client)
    }
    guard status == NativeConstants.statusOK else {
      throw error(for: status)
    }
    guard let client else {
      throw CaptureError.incompatibleNative(reason: "native create returned a null client")
    }
    return client
  }

  internal static func error(for status: UInt32) -> CaptureError {
    switch status {
    case NativeConstants.statusInvalidArgument,
      NativeConstants.statusInvalidStructSize:
      .invalidArgument(status: status)
    case NativeConstants.statusClientClosed:
      .clientClosed
    case NativeConstants.statusOutOfMemory:
      .outOfMemory
    default:
      .nativeStatus(rawValue: status)
    }
  }
}

private let captureCallback: NativeCaptureCallback = { _, completion, userData in
  guard let userData else {
    return
  }
  let box = Unmanaged<RequestBox>.fromOpaque(userData).takeRetainedValue()
  box.complete(completion: completion)
}

private final class ClientState: @unchecked Sendable {
  private let lock = NSLock()
  fileprivate let native: any NativeCaptureAPI
  private var client: OpaquePointer?

  fileprivate init(native: any NativeCaptureAPI, client: OpaquePointer) {
    self.native = native
    self.client = client
  }

  fileprivate func start(
    options: UnsafePointer<snaploom_capture_options_v1>,
    callback: NativeCaptureCallback,
    userData: UnsafeMutableRawPointer,
    requestID: UnsafeMutablePointer<UInt64>
  ) -> UInt32 {
    lock.lock()
    defer { lock.unlock() }
    guard let client else {
      return NativeConstants.statusClientClosed
    }
    return native.start(client, options, callback, userData, requestID)
  }

  fileprivate func requestCancel(requestID: UInt64) {
    lock.lock()
    defer { lock.unlock() }
    guard let client else {
      return
    }
    _ = native.cancel(client, requestID)
  }

  fileprivate func close() throws {
    let closing: OpaquePointer?
    lock.lock()
    closing = client
    client = nil
    lock.unlock()
    guard let closing else {
      return
    }
    let status = native.destroy(closing)
    guard status == NativeConstants.statusOK else {
      lock.lock()
      if client == nil {
        client = closing
      }
      lock.unlock()
      throw CaptureClient.error(for: status)
    }
  }
}

private final class RequestBox: @unchecked Sendable {
  private let lock = NSLock()
  private let native: any NativeCaptureAPI
  private let cancel: @Sendable (UInt64) -> Void
  private var continuation: CheckedContinuation<CaptureResult, any Error>?
  private var requestID: UInt64?
  private var cancelRequested = false
  private var terminal = false

  fileprivate init(
    native: any NativeCaptureAPI,
    cancel: @escaping @Sendable (UInt64) -> Void
  ) {
    self.native = native
    self.cancel = cancel
  }

  fileprivate func install(
    continuation: CheckedContinuation<CaptureResult, any Error>
  ) {
    lock.lock()
    self.continuation = continuation
    lock.unlock()
  }

  fileprivate func publish(requestID: UInt64) {
    let shouldCancel: Bool
    lock.lock()
    if terminal {
      lock.unlock()
      return
    }
    self.requestID = requestID
    shouldCancel = cancelRequested
    lock.unlock()
    if shouldCancel {
      cancel(requestID)
    }
  }

  fileprivate func requestCancel() {
    let published: UInt64?
    lock.lock()
    cancelRequested = true
    published = terminal ? nil : requestID
    lock.unlock()
    if let published {
      cancel(published)
    }
  }

  fileprivate func failStart(error: CaptureError) {
    let waiting = finish()
    waiting.continuation?.resume(throwing: error)
  }

  fileprivate func complete(
    completion: UnsafeMutablePointer<snaploom_capture_completion_v1>?
  ) {
    guard let completion else {
      let waiting = finish()
      waiting.continuation?.resume(
        throwing: CaptureError.invalidCompletion(reason: "native completion is null")
      )
      return
    }
    defer { native.freeCompletion(completion) }

    let waiting = finish()
    guard let continuation = waiting.continuation else {
      return
    }
    do {
      continuation.resume(
        returning: try decode(
          completion: completion.pointee,
          callerCanceled: waiting.cancelRequested
        )
      )
    } catch {
      continuation.resume(throwing: error)
    }
  }

  private func finish() -> (
    continuation: CheckedContinuation<CaptureResult, any Error>?,
    cancelRequested: Bool
  ) {
    lock.lock()
    defer { lock.unlock() }
    guard !terminal else {
      return (nil, cancelRequested)
    }
    terminal = true
    let waiting = continuation
    continuation = nil
    return (waiting, cancelRequested)
  }

  private func decode(
    completion: snaploom_capture_completion_v1,
    callerCanceled: Bool
  ) throws -> CaptureResult {
    guard completion.struct_size >= MemoryLayout<snaploom_capture_completion_v1>.size else {
      throw CaptureError.invalidCompletion(reason: "native completion layout is too short")
    }
    switch completion.kind {
    case NativeConstants.completionCompleted:
      guard
        completion.png_size > 0,
        completion.png_size <= NativeConstants.maximumPNGSize,
        let png = completion.png_data,
        let count = Int(exactly: completion.png_size),
        completion.pixel_width > 0,
        completion.pixel_height > 0
      else {
        throw CaptureError.invalidCompletion(reason: "completed PNG metadata is invalid")
      }
      return CaptureResult(
        pngData: Data(bytes: png, count: count),
        pixelWidth: completion.pixel_width,
        pixelHeight: completion.pixel_height,
        clipboardWritten: completion.flags & NativeConstants.clipboardWritten != 0
      )
    case NativeConstants.completionCanceled:
      if callerCanceled {
        throw CancellationError()
      }
      throw CaptureError.canceledByUser
    case NativeConstants.completionFailed:
      throw CaptureError.failure(
        CaptureFailure(
          code: CaptureErrorCode(rawValue: completion.error_code),
          isRetryable: completion.flags & NativeConstants.retryable != 0
        )
      )
    default:
      throw CaptureError.unknownCompletion(rawValue: completion.kind)
    }
  }
}
